### ISSUE TITLE
[miniplex-react] Updating Component data prop results in the owning entity being deleted and re-added

### ISSUE THREAD

**Issue by madou** [hmans/miniplex#172]
Package: `miniplex-react@1.0.0`
Sandbox: https://codesandbox.io/s/entity-deleted-when-updating-component-cmk0og?file=/src/App.tsx

---

When re-rendering a Component component it calls `world.removeComponent` and then `world.addComponent` in an effect. During which it unexpectedly, at least as far as the signals are considered, deleting and then adding the entity to the world. This proves to be very bad when using the `useArchetype` hook as it re-renders _a lot_.

This also seems counter productive as I wouldn't have expected the component to be replaced, but instead mutated, after the first render. Modifying the code to this locally to set on initial render and mutate on subsequent ones fixes the problem but since it's replacing the reference it might be doing something dangerous [according to the docs](https://github.com/hmans/miniplex/tree/main/packages/miniplex#do-not-add-or-remove-entity-properties-directly):

```js
useIsomorphicLayoutEffect(() => {
  if (name in entity) {
    entity[name] = data;
  } else {
    world.addComponent(entity, name, data ?? (ref.current as any));
  }
}, [entity, name, data]);

useIsomorphicLayoutEffect(() => {
  return () => {
    if ("__miniplex" in entity) {
      world.removeComponent(entity, name);
    }
  };
}, [name, entity]);

```

There may be a bug in the achetype indexing. Regardless, I'd imagine skipping the need for indexing could be beneficial.

**Comment by hmans** [comment]
Hi again, and thanks for submitting this issue!

I agree that `<Component>` should not _remove_ and then immediately _re-add_ the component and would consider this a bug, and I will look into it.

I am not 100% sure about having it update the component data on re-renders, at least not by default, since this would interfere with imperative systems mutating the data, and probably be a major breaking change (compare this to, eg. creating a scene object in R3F with an initial position, where you typically don't want the scene object to reset to that position every time the component re-renders.)

One option here would be to introduce an `update` boolean prop to `<Component>` that will update the component with the `data` or `ref` every time it is re-rendered. What do you think?

**Comment by hmans** [comment]
Addendum: since archetype indices only care about the _presence_ of a component (and not its value), the pattern that you implemented above is perfectly fine, since the _presence_ of the component is not affected. The bugfix/change I will be implementing will be doing a very similar thing.

**Comment by hmans** [comment]
> I am not 100% sure about having it update the component data on re-renders, at least not by default, since this would interfere with imperative systems mutating the data, and probably be a major breaking change (compare this to, eg. creating a scene object in R3F with an initial position, where you typically don't want the scene object to reset to that position every time the component re-renders.)

I'm going to self-reply here: no, I was wrong, `<Component>` is actually designed to work this way (updating the component on re-renders), it just does it in a dangerous way that needs fixing. :-)

**Comment by madou** [comment]
Cool thanks mate! Id say it would be great if component has an initialData prop for scenarios where you only want to set it once. I've found myself setting it with a cached value that never changes to get the behaviour.

And since I'm here one other aspect I got caught out with is the data value passed in is the one that is eventually mutated. So I by accident had all my entities sharing a velocity component based on my "empty" shared value.
